### PR TITLE
feat(git): insteadOf environment variables for authentication

### LINKED ISSUE #11060: Refactor: use env to pass credentials to go get
Broken off from #7361 for @Shegox

### PR DESCRIPTION
**Pull request by Shegox**
## Changes:

This PR uses the newly in git 2.31 introduced environment variables for authentication for go module updates (e.g. sum files). 
The current change only replaces the existing functionality and provides a generic function to get authentication environment variables for git via `getGitAuthenticatedEnvironmentVariables(gitUrl, token, environmentVariables)`. 

Requires git cli [>=2.31](https://github.com/git/git/blob/master/Documentation/RelNotes/2.31.0.txt) to use [`GIT_CONFIG_` variables](https://git-scm.com/docs/git-config#Documentation/git-config.txt-GITCONFIGCOUNT) to provide the credentials as environment variables to the `go get` process, which then invokes `git` underneath.

Happy about any feedback :) 

## Context:

Fixes #11060

Part of enabling authenticated go module updates #7361 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/auth.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    gitConfigCount = parseInt(gitConfigCountEnvVariable, 10);
+    if (Number.isNaN(gitConfigCount)) {
+      logger.warn(
+        `Found GIT_CONFIG_COUNT env variable, but couldn't parse the value to an integer: ${process.env.GIT_CONFIG_COUNT}. Ignoring it.`
+      );
+      gitConfigCount = 0;
+    }
+  }
+
+  const gitUrlWithToken = getHttpUrl(gitUrl, encodeURIComponent(token));
```
Comment (viceice) [inline lib/util/git/auth.ts:27]: Are you sure to encode the token here?
```suggestion
  const gitUrlWithToken = getHttpUrl(gitUrl, token);
```

I think `getHttpUrl` is already doing this for us.

**Comment by viceice** [comment]
BTW: git v 2.31.0 is from march 2021. so we should add a hint for selfhosted users

https://github.com/renovatebot/renovate/blob/8fac1fdeff5a74334fa1e3e4e16b1e321aaaf0e4/docs/development/local-development.md#L13-L13

https://github.com/renovatebot/renovate/blob/36d37b328147ace4df91b21e90d62d3c451e5c51/docs/usage/self-hosting.md#L1-L1

File: lib/util/git/auth.ts (hunk 1 above)
Comment (Shegox) [reply lib/util/git/auth.ts:27]: Not as far as I can see it, `getRemoteUrlWithToken()` as  well uses the same syntax: 
https://github.com/renovatebot/renovate/blob/3ffb57b1e8fa17360f1caf6c1c5fae97246f2a7a/lib/util/git/url.ts#L22

But I think we can remove the quote around the token in the `artifact.ts`:
https://github.com/renovatebot/renovate/blob/3ffb57b1e8fa17360f1caf6c1c5fae97246f2a7a/lib/manager/gomod/artifacts.ts#L30

Comment (viceice) [reply lib/util/git/auth.ts:27]: Can we verify that L13 did not already encode the token:
https://github.com/renovatebot/renovate/blob/38f8b55c738fb38c926bef8b716fe97eba105254/lib/util/git/url.ts#L5-L14

I also think github tokens don't need to be encoded at all.

Comment (Shegox) [reply lib/util/git/auth.ts:27]: Line 13 does not correctly encode it. 
When trying it without the encoding it ended up looking like this, without the encoded `:`.
```
    -   "GIT_CONFIG_KEY_0": "url.https://token%3A1234@github.com/.insteadOf",
    +   "GIT_CONFIG_KEY_0": "url.https://token:1234@github.com/.insteadOf",
```
I added as well a test to ensure that it is correctly encoded.

**Comment by rarkins** [comment]
I tried adding a git version detection previously and had to give up. Surely shouldn't be too hard. Maybe we could try with v26 and have an option to skip detection if it fails?

**Comment by Shegox** [comment]
Yes, that could be an option. The worst happening is that git ignores the environment variables. Meaning that just the authentication for`go get -d` fails, everything else keeps on working as currently. 
But yeah it makes sense to at least log an error/warning message if it isn't supported.

**Comment by viceice** [comment]
> I tried adding a git version detection previously and had to give up. Surely shouldn't be too hard. Maybe we could try with v26 and have an option to skip detection if it fails?

I know, it's on my to-do list. Will check what's going wrong on your branch.

**Comment by rarkins** [comment]
> Yes, that could be an option. The worst happening is that git ignores the environment variables. Meaning that just the authentication for`go get -d` fails, everything else keeps on working as currently.
> But yeah it makes sense to at least log an error/warning message if it isn't supported.

Because the failure is therefore graceful, I'm ok with merging this without adding logic to check/warn about git, because that's not for go modules only.

However if Go Modules auth was working before with a certain version of git but then stop working in a newer version of Renovate, I think that's probably best to go into v26 as a breaking change so that it's less likely to be missed.

**Comment by Shegox** [comment]
@rarkins and @viceice I took the time to rebase my changes. 
Since nowadays git `>=2.33` is anyway required this (requiring git `>=2.31`) should no longer be a breaking change and good to be reviewed.

**Review by viceice — APPROVED** [review]

**Pull request merged**